### PR TITLE
Block fetch inflight bug

### PR DESCRIPTION
**Pull request by coot**
I tried to construct a test case which catches the bug (#973), but it requires a bit more work. I think that the current block fetch test never merges requested chains.

### CODE REVIEW

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs [hunk 1]
```diff
@@ -357,6 +364,85 @@ tracePropertyClientStateSanity es =
...
+  where
+    f (TraceFetchClientState (TraceLabelPeer peerid (AddedFetchRequest request inflight _ _))) =
+      Just (TraceLabelPeer peerid (AddedFetchRequestTrace request inflight))
+    f (TraceFetchClientState (TraceLabelPeer peerid (AcknowledgedFetchRequest{}))) =
+      Just (TraceLabelPeer peerid AcknowledgedFetchRequestTrace)
+    f (TraceFetchClientState (TraceLabelPeer peerid CompletedFetchBatch))
+      = Just (TraceLabelPeer peerid CompletedFetchBatchTrace)
+    f _ = Nothing
+
+tracePropertyInFlight :: [Example1TraceEvent] -> Property
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs]: Done.

File: ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs [hunk 2]
```diff
@@ -357,6 +364,85 @@ tracePropertyClientStateSanity es =
...
+  where
+    checkTrace :: Maybe (FetchRequest BlockHeader)
+               --  acknowledged 'FetchRequest'
+               -> Int
+               -- number of requests in flight (not accounting the above
+               -- acknowledged 'FetchRequest')
+               -> [FetchRequestTrace]
+               -> Property
+
+    -- AddedFetchRequest: check that the number of requests in flights.
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/test/Test/Ouroboros/Network/BlockFetch.hs]: Yes; I rewrote the comments, some of them were not accurate.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 3]
```diff
@@ -428,13 +444,21 @@ type TFetchRequestVar m header =
 newTFetchRequestVar :: MonadSTM m => STM m (TFetchRequestVar m header)
 newTFetchRequestVar = newTMergeVar
 
+
+-- | Write to the underlyaing 'TMergeVar' and return the number of new
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Fixed

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs [hunk 4]
```diff
@@ -462,13 +485,20 @@ newtype TMergeVar m a = TMergeVar (StrictTMVar m a)
...
 newTMergeVar = TMergeVar <$> newEmptyTMVar
 
-writeTMergeVar :: (MonadSTM m, Semigroup a) => TMergeVar m a -> a -> STM m ()
+-- | Merge the current value with the given one and store it, return the updated
+-- value.
+--
+writeTMergeVar :: (MonadSTM m, Semigroup a) => TMergeVar m a -> a -> STM m a
 writeTMergeVar (TMergeVar v) x = do
     mx0 <- tryTakeTMVar v
     case mx0 of
```
[the remark this answers is not in the document]
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: That's not possible, different branches use different value.

**Comment by coot** [comment]
I hopefully this time hydra will not abort.

File: ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs (hunk 4 above)
[the remark this answers is not in the document]
Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs]: Just FYI, no need to use this, but this is the direction I had in mind.

```
let x' = case mx0 of
      Nothing -> x
      Just x0 -> x0 <> x
x' <$ putTMVar x'
```

**Pull request merged**